asllib/Interpreter.ml

```diff
@@ -485,6 +470,68 @@ module Make (B : Backend.S) (C : Config) = struct
         in
         fatal_from e (Error.UnexpectedSideEffect msg)
 
+  (* Runtime checks *)
+  (* -------------- *)
+
+  and is_val_of_type loc env v ty : bool B.m =
+    let m_true = L_Bool true |> B.v_of_literal |> return in
+    let m_false = L_Bool false |> B.v_of_literal |> return in
+    let and' prev here = prev >>= B.binop BAND here in
+    let or' prev here = prev >>= B.binop BOR here in
+    let rec in_values v ty =
+      match (Types.get_structure (IEnv.to_static env) ty).desc with
+      | T_Real | T_Bool | T_Enum _ | T_String | T_Bits _ | T_Int None -> m_true
+      | T_Int (Some constraints) ->
+          let fold prev = function
+            | Constraint_Exact e ->
+                let* v' = eval_expr_sef env e in
+                let* here = B.binop EQ_OP v v' in
+                or' prev here
+            | Constraint_Range (e1, e2) ->
+                let* v1 = eval_expr_sef env e1 and* v2 = eval_expr_sef env e2 in
+                let* here =
+                  let* c1 = B.binop LEQ v1 v and* c2 = B.binop LEQ v v2 in
+                  B.binop BAND c1 c2
+                in
+                or' prev here
+          in
+          List.fold_left fold m_false constraints
+      | T_Record fields | T_Exception fields ->
+          let fold prev (field_name, field_type) =
+            let* v' = B.get_field field_name v in
+            let* here = in_values v' field_type in
+            and' prev here
+          in
+          List.fold_left fold m_true fields
+      | T_Tuple tys ->
+          let fold (i, prev) ty' =
+            let m =
+              let* v' = B.get_index i v in
+              let* here = in_values v' ty' in
+              and' prev here
+            and i = i + 1 in
+            (i, m)
+          in
+          List.fold_left fold (0, m_true) tys |> snd
+      | T_Array (e, ty') ->
+          let* v = eval_expr_sef env e in
+          let n =
+            match B.v_to_int v with
+            | Some i -> i
+            | None -> fatal_from loc @@ Error.UnsupportedExpr e
+          in
+          let rec loop i prev =
+            if i = n then prev
+            else
+              let* v' = B.get_index i v in
+              let* here = in_values v' ty' in
+              loop (succ i) (and' prev here)
+          in
+          loop 0 m_true
+      | T_Named _ -> assert false
+    in
+    B.choice (in_values v ty) (return true) (return false)
```

Comment:
This function is unused...

Reply:
Good catch! I somehow managed not to commit it...

It should be fixed with the last pushed version.

I also integrated your example, see the error message in `asllib/tests/asl.t/run.t`. Do you think I should test more on this feature?